src/phasorpy/_phasorpy.pyx

```diff
@@ -2034,3 +2036,106 @@ def _signal_denoise_vector(
             n = <float_t> (<double> integrated[i] / sum)
             for m in range(samples):
                 denoised[i, m] *= n
+
+
+###############################################################################
+# Filtering functions
+
+cdef float_t _quickselect(
+    float_t *arr,
+    ssize_t left,
+    ssize_t right,
+    const ssize_t k
+) noexcept nogil:
+    """Quickselect algorithm to find the k-th smallest element."""
+    cdef:
+        ssize_t i, pivot_index, pivot_new_index
+        float_t pivot_value, temp
+
+    while left <= right:
+        pivot_index = left + (right - left) // 2
+        pivot_value = arr[pivot_index]
+        temp = arr[pivot_index]
+        arr[pivot_index] = arr[right]
+        arr[right] = temp
+        pivot_new_index = left
+        for i in range(left, right):
+            if arr[i] < pivot_value:
+                temp = arr[i]
+                arr[i] = arr[pivot_new_index]
+                arr[pivot_new_index] = temp
+                pivot_new_index += 1
+        temp = arr[right]
+        arr[right] = arr[pivot_new_index]
+        arr[pivot_new_index] = temp
+
+        if pivot_new_index == k:
+            return arr[k]
+        if pivot_new_index < k:
+            left = pivot_new_index + 1
+        else:
+            right = pivot_new_index - 1
+
+    return arr[k]
+
+
+cdef float_t _median(const float_t *values, const ssize_t n) noexcept nogil:
+    """Calculate the median of an array of values."""
+    if n % 2 == 0:
+        return (
+            (_quickselect(values, 0, n - 1, n // 2 - 1) +
+             _quickselect(values, 0, n - 1, n // 2)) / 2
+        )
+
+    return _quickselect(values, 0, n - 1, n // 2)
+
+
+def _apply_2d_median_filter(
+    const float_t[:, :] image,
+    float_t[:, :] filtered_image,
```

Comment:
As far as I can see, you always pass C contiguous array copies to this function. Then this should be faster:
```
    const float_t[:, ::1] image,
    float_t[:, ::1] filtered_image,  # copy of image
```

Reply:
Thanks Christoph for this suggestion! This helped a little bit and now the Cython implementation is around 10-15% faster than Scipy's for 2D filtering.